src/algorithms/calorimetry/ImagingTopoCluster.h

```diff
@@ -124,21 +125,61 @@ namespace eicrecon {
 
-        // group neighbouring hits
-        std::vector<bool> visits(hits->size(), false);
-        std::vector<std::set<std::size_t>> groups;
-        for (size_t i = 0; i < hits->size(); ++i) {
-            debug("hit {:d}: local position = ({}, {}, {}), global position = ({}, {}, {})", i + 1,
-                         (*hits)[i].getLocal().x, (*hits)[i].getLocal().y, (*hits)[i].getPosition().z,
-                         (*hits)[i].getPosition().x, (*hits)[i].getPosition().y, (*hits)[i].getPosition().z
+        // Sort hit indices (podio collections do not support std::sort)
+        auto compare = [&hits](const auto& a, const auto& b) {
+            // if !(a < b) and !(b < a), then a and b are equivalent
+            // and only one of them will be allowed in a set
+            if ((*hits)[a].getLayer() == (*hits)[b].getLayer()) {
+              return (*hits)[a].getObjectID().index < (*hits)[b].getObjectID().index;
+            }
+            return (*hits)[a].getLayer() < (*hits)[b].getLayer();
+        };
+        // indices contains the remaining hit indices that have not
+        // been assigned to a group yet
+        std::set<std::size_t, decltype(compare)> indices(compare);
+        // set does not have a size yet, so cannot fill with iota
+        for (std::size_t i = 0; i < hits->size(); ++i) {
+            indices.insert(i);
+        }
+        // ensure no hits were dropped due to equivalency in set
+        if (hits->size() != indices.size()) {
+            error("equivalent hits were dropped: #hits {:d}, #indices {:d}", hits->size(), indices.size());
+        }
+
+        // Group neighbouring hits
+        std::vector<std::list<std::size_t>> groups;
+        // because indices changes, the loop over indices requires some care:
+        // - we must use iterators instead of range-for
+        // - erase returns an incremented iterator and therefore acts as idx++
+        // - when the set becomes empty on erase, idx is invalid and idx++ will be too
+        // (also applies to loop in bfs_group below)
+        for (auto idx = indices.begin(); idx != indices.end();
+                 indices.empty() ? idx = indices.end() : idx) {
+
+            debug("hit {:d}: local position = ({}, {}, {}), global position = ({}, {}, {}), energy = {}", *idx,
+                         (*hits)[*idx].getLocal().x, (*hits)[*idx].getLocal().y, (*hits)[*idx].getPosition().z,
+                         (*hits)[*idx].getPosition().x, (*hits)[*idx].getPosition().y, (*hits)[*idx].getPosition().z,
+                         (*hits)[*idx].getEnergy()
             );
-            // already in a group, or not energetic enough to form a cluster
-            if (visits[i] || (*hits)[i].getEnergy() < minClusterCenterEdep) {
-                continue;
+
+            // not energetic enough for cluster center or other cluster hits, so erase
+            if ((*hits)[*idx].getEnergy() < std::min(m_cfg.minClusterHitEdep,m_cfg.minClusterCenterEdep)) {
```

Comment:
I propose dropping this check right here (as I explain below). The first thing bfs_group will do is repeat this check+erasure. That way there are less preconditions to keep in mind in bfs_group, and less code duplication.

Reply:
3d151e22a